ksql-rest-app/src/test/java/io/confluent/ksql/rest/server/computation/SequenceNumberFutureStoreTest.java

```diff
@@ -0,0 +1,101 @@
+/*
+ * Copyright 2018 Confluent Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.confluent.ksql.rest.server.computation;
+
+import static org.hamcrest.CoreMatchers.is;
+import static org.hamcrest.CoreMatchers.not;
+import static org.hamcrest.CoreMatchers.sameInstance;
+import static org.hamcrest.MatcherAssert.assertThat;
+
+import java.util.concurrent.CompletableFuture;
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.ExpectedException;
+
+public class SequenceNumberFutureStoreTest {
+
+  @Rule
+  public final ExpectedException expectedException = ExpectedException.none();
+
+  private SequenceNumberFutureStore sequenceNumberFutureStore;
+
+  @Before
+  public void setUp() {
+    sequenceNumberFutureStore = new SequenceNumberFutureStore();
+  }
+
+  @Test
+  public void shouldReturnFutureForNewSequenceNumber() {
+    // When:
+    final CompletableFuture<Void> future = sequenceNumberFutureStore.getFutureForSequenceNumber(2);
+
+    // Then:
+    assertFutureIsNotCompleted(future);
+  }
+
+  @Test
+  public void shouldReturnFutureForExistingSequenceNumber() {
+    // Given:
+    final CompletableFuture<Void> existingFuture = sequenceNumberFutureStore.getFutureForSequenceNumber(2);
+
+    // When:
+    final CompletableFuture<Void> newFuture = sequenceNumberFutureStore.getFutureForSequenceNumber(2);
+
+    // Then:
+    assertThat(newFuture, is(sameInstance(existingFuture)));
+  }
+
+  @Test
+  public void shouldReturnFutureForCompletedSequenceNumber() {
+    // Given:
+    final CompletableFuture<Void> firstFuture = sequenceNumberFutureStore.getFutureForSequenceNumber(2);
+    sequenceNumberFutureStore.completeFuturesUpThroughSequenceNumber(2);
+    assertFutureIsCompleted(firstFuture);
+
+    // When:
+    final CompletableFuture<Void> secondFuture = sequenceNumberFutureStore.getFutureForSequenceNumber(2);
+
+    // Then:
+    assertFutureIsCompleted(secondFuture);
+    assertThat(secondFuture, is(not(sameInstance(firstFuture))));
+  }
+
+  @Test
+  public void shouldCompleteFutures() {
+    // Given:
+    final CompletableFuture<Void> firstFuture = sequenceNumberFutureStore.getFutureForSequenceNumber(2);
+    final CompletableFuture<Void> secondFuture = sequenceNumberFutureStore.getFutureForSequenceNumber(3);
+
+    // When:
+    sequenceNumberFutureStore.completeFuturesUpThroughSequenceNumber(2);
+
+    // Then:
+    assertFutureIsCompleted(firstFuture);
```

Comment:
FYI, rather than such methods you can create reusable Matchers that allow you to write code like:

```java
assertThat(firstFuture, is(completedNormally()));
assertThat(secondFurture, is(not(complete())));
```

e.g.

```java
public final class CompleteableFutureMatchers {
   public static Matcher<? super CompletableFuture<?>> completedNormally() {
    return new TypeSafeDiagnosingMatcher<CompletableFuture<?>>() {
      @Override
      protected boolean matchesSafely(final CompletableFuture<?> f, final Description mismatchDescription) {
        if (f.isCancelled()) {
          mismatchDescription.appendText("the future was cancelled");
          return false;
        }

        if (f.isCompletedExceptionally()) {
          mismatchDescription.appendText("the future completed exceptionally");
          return false;
        }

        if f.isDone()) {
          mismatchDescription.appendText("the future is not complete");
          return false;
        }

        return true;
      }

      @Override
      public void describeTo(final Description description) {
        description.appendText("a future that completed normally");
      }
    };
  }
}
```

Which leads to easier to read tests.